src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,282 @@
+                RuleLevel.BuildWarning,
+                CreateLocalizableResourceString(nameof(FeatureUnsupportedWhenRuntimeMarshallingDisabledDescription)),
+                isPortedFxCopRule: false,
+                isDataflowRule: false);
+
+        internal const string MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabledId = "CA1421";
+
+        private static readonly DiagnosticDescriptor MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabled =
+            DiagnosticDescriptorHelper.Create(
+                MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabledId,
+                CreateLocalizableResourceString(nameof(MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabledTitle)),
+                CreateLocalizableResourceString(nameof(MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabledMessage)),
+                DiagnosticCategory.Interoperability,
+                RuleLevel.IdeSuggestion,
+                CreateLocalizableResourceString(nameof(MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabledDescription)),
+                isPortedFxCopRule: false,
+                isDataflowRule: false);
+
+        public const string CanConvertToDisabledMarshallingEquivalentKey = nameof(CanConvertToDisabledMarshallingEquivalentKey);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            FeatureUnsupportedWhenRuntimeMarshallingDisabled,
+            MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabled);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (context.Compilation.TryGetOrCreateTypeByMetadataName(
+                    WellKnownTypeNames.SystemRuntimeCompilerServicesDisableRuntimeMarshallingAttribute,
+                    out INamedTypeSymbol? disableRuntimeMarshallingAttribute)
+                    && context.Compilation.Assembly.HasAttribute(disableRuntimeMarshallingAttribute))
+                {
+                    var perCompilationAnalyzer = new PerCompilationAnalyzer(context.Compilation);
+                    perCompilationAnalyzer.RegisterActions(context);
+                }
+            });
+        }
+
+        private class PerCompilationAnalyzer
+        {
+            private readonly INamedTypeSymbol? _unmanagedFunctionPointerAttribute;
+            private readonly INamedTypeSymbol? _structLayoutAttribute;
+            private readonly INamedTypeSymbol? _lcidConversionAttribute;
+            private readonly ImmutableArray<ISymbol> _marshalMethods;
+            private readonly ReaderWriterLockSlim _layoutCacheLock = new();
+            private readonly Dictionary<ITypeSymbol, bool> _isAutoLayoutOrContainsAutoLayoutCache = new();
```

Comment:
We normally use a `ConcurrentDictionary` in this repo, but this is fine if you prefer dictionary guarded with a lock. Is there a specific reason for using it here or just your preference? try/catch/reader writer locks do make the code look bit more verbose

Reply:
ConcurrentDictionary has a lot of built-in overhead, so I’m a little weary of using it by default without benchmarking (I’ve been burned by perf issues from using it).

Reply:
@jkoritzinsky, can you elaborate on what overhead you're concerned about? RWLS generally has way more overhead.

Reply:
When I was working on C#/WinRT, I defaulted to using ConcurrentDictionary when I needed a multithreaded dictionary, and it’s mechanisms of allocating caches across multiple threads to ensure efficiency has caused no end of perf headaches due to how many ConcurrentDictionary instances are used there. If this isn’t a concern, I can switch this back to ConcurrentDictionary.

Reply:
I don't believe it should be a concern until you measure it to be a concern (and if perf is the driving reason, you also need to measure the overhead of RWLS, which has way higher overhead on every read and write).  But if it's really a concern, you can set the concurrencyLevel to 1 in the ctor.